File: app/src/main/java/com/defianttech/convertme/ConvertActivity.java
package com.defianttech.convertme;

import java.text.DecimalFormat;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewCompat;
import android.support.v4.widget.Space;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.view.ActionMode;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.Toolbar;
import android.text.Html;
import android.text.Spanned;
import android.util.Log;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

/*
 * Copyright (c) 2014-2017 Dmitry Brant
 */
public class ConvertActivity extends AppCompatActivity {
    private static final String TAG = "ConvertActivity";
    private static final String PREFS_NAME = "ConvertMePrefs";
    private static final String KEY_CURRENT_CATEGORY = "currentCategory";
    private static final String KEY_CURRENT_UNIT = "currentUnitIndex";
    private static final String KEY_CURRENT_VALUE = "currentValue";

    private static DecimalFormat dfExp = new DecimalFormat("#.#######E0");
    private static DecimalFormat dfNoexp = new DecimalFormat("#.#######");

    private UnitCollection[] collections;
    private String[] allCategoryNames;

    private int currentCategory = UnitCollection.DEFAULT_CATEGORY;
    private int currentUnitIndex = UnitCollection.DEFAULT_FROM_INDEX;

    private double currentValue = UnitCollection.DEFAULT_VALUE;

    private TextView categoryText;
    private PopupMenu categoryMenu;

    private NumberPadView numberPadView;
    private UnitListAdapter listAdapter;
    private ListView unitsList;
    private FloatingActionButton fabEdit;
    private ActionMode actionMode;
    private boolean editModeEnabled;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.convertme);

        collections = UnitCollection.getInstance(this);
        allCategoryNames = UnitCollection.getAllCategoryNames(this);

        Toolbar toolbar = findViewById(R.id.main_toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayShowTitleEnabled(false);
        }

        View categoryContainer = findViewById(R.id.category_toolbar_container);
        categoryContainer.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                categoryMenu.show();
            }
        });

        categoryMenu = new PopupMenu(ConvertActivity.this, categoryContainer);
        int i = 0;
        for (String name : allCategoryNames) {
            categoryMenu.getMenu().add(0, i++, 0, name);
        }
        categoryMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                currentCategory = UnitCollection.collectionIndexByName(collections,
                        allCategoryNames[item.getItemId()]);
                if (currentUnitIndex >= collections[currentCategory].length()) {
                    currentUnitIndex = 0;
                }
                categoryText.setText(item.getTitle());
                listAdapter.notifyDataSetInvalidated();
                return true;
            }
        });

        categoryText = findViewById(R.id.category_text);
        unitsList = findViewById(R.id.unitsList);
        listAdapter = new UnitListAdapter();
        unitsList.setAdapter(listAdapter);
        unitsList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                if (editModeEnabled) {
                    collections[currentCategory].get(position)
                            .setEnabled(!collections[currentCategory].get(position).isEnabled());
                } else {
                    currentUnitIndex = position;
                }
                listAdapter.notifyDataSetChanged();
            }
        });
        unitsList.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> adapterView, View view, int position, long id) {
                if (editModeEnabled) {
                    return false;
                }
                doLongPressMenu(view.findViewById(R.id.unitValue), position);
                return true;
            }
        });

        numberPadView = findViewById(R.id.numberPad);
        numberPadView.setOnValueChangedListener(new NumberPadView.OnValueChangedListener() {
            @Override
            public void onValueChanged(String value) {
                setValueFromNumberPad(value);
                listAdapter.notifyDataSetChanged();
            }
        });

        restoreSettings();

        for (String name : allCategoryNames) {
            if (name.equals(collections[currentCategory].getNames()[0])) {
                categoryText.setText(name);
            }
        }

        fabEdit = findViewById(R.id.fabEdit);
        fabEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startSupportActionMode(new EditUnitsActionModeCallback());
            }
        });
    }

    @Override
    public void onStop(){
        super.onStop();
        saveSettings();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu){
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item){
        switch(item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
            case R.id.menu_about:
                showAboutDialog();
                return true;
            default:
                break;
        }
        return false;
    }

    @NonNull
    public static SharedPreferences getPrefs(@NonNull Context context) {
        return context.getSharedPreferences(ConvertActivity.PREFS_NAME, 0);
    }

    private void saveSettings() {
        SharedPreferences.Editor editor = getPrefs(this).edit();
        editor.putInt(KEY_CURRENT_CATEGORY, currentCategory);
        editor.putInt(KEY_CURRENT_UNIT, currentUnitIndex);
        editor.putString(KEY_CURRENT_VALUE, numberPadView.getCurrentValue());
        for (UnitCollection col : collections) {
            for (SingleUnit unit : col.getItems()) {
                editor.putBoolean(unit.getName(), unit.isEnabled());
            }
        }
        editor.apply();
    }

    private void restoreSettings() {
        try {
            SharedPreferences prefs = getPrefs(this);
            currentCategory = prefs.getInt(KEY_CURRENT_CATEGORY, UnitCollection.DEFAULT_CATEGORY);
            if (currentCategory >= collections.length) {
                currentCategory = UnitCollection.DEFAULT_CATEGORY;
            }
            currentUnitIndex = prefs.getInt(KEY_CURRENT_UNIT, UnitCollection.DEFAULT_FROM_INDEX);
            if (currentUnitIndex >= collections[currentCategory].length()) {
                currentUnitIndex = 0;
            }
            numberPadView.setCurrentValue(prefs.getString(KEY_CURRENT_VALUE, "1"));
            setValueFromNumberPad(numberPadView.getCurrentValue());
            for (UnitCollection col : collections) {
                for (SingleUnit unit : col.getItems()) {
                    unit.setEnabled(prefs.getBoolean(unit.getName(), true));
                }
            }
        } catch(Exception ex) {
            //ehh...
        }
    }

    private void updateActionModeState() {
        numberPadView.setVisibility(editModeEnabled ? View.GONE : View.VISIBLE);
        if (editModeEnabled) {
            fabEdit.hide();
        } else {
            fabEdit.show();
        }
        listAdapter.notifyDataSetInvalidated();
    }

    private void setValueFromNumberPad(@NonNull String value) {
        try {
            currentValue = Double.parseDouble(value);
        } catch (NumberFormatException e) {
            currentValue = 0.0;
        }
    }

    private final class EditUnitsActionModeCallback implements ActionMode.Callback {
        @ColorInt int statusBarColor;

        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            actionMode = mode;
            actionMode.setTitle(getString(R.string.show_hide_units));
            editModeEnabled = true;
            updateActionModeState();
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                statusBarColor = getWindow().getStatusBarColor();
                getWindow().setStatusBarColor(Color.BLACK);
            }
            return false;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            if (item.getItemId() == android.R.id.home) {
                mode.finish();
                return true;
            }
            return false;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                getWindow().setStatusBarColor(statusBarColor);
            }
            actionMode = null;
            editModeEnabled = false;
            updateActionModeState();
        }
    }

    private final class UnitListAdapter extends BaseAdapter {
        @Override
        public int getCount() {
            return collections[currentCategory].length();
        }

        @Override
        public Object getItem(int position) {
            return collections[currentCategory].get(position);
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public int getItemViewType(int position) {
            if (editModeEnabled) {
                return 0;
            } else {
                return collections[currentCategory].get(position).isEnabled() ? 0 : 1;
            }
        }

        @Override
        public int getViewTypeCount() {
            return 2;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            if (editModeEnabled) {

                if (convertView == null) {
                    convertView = getLayoutInflater().inflate(R.layout.unit_listitem, parent, false);
                }
                View itemContainer = convertView.findViewById(R.id.unitItemContainer);
                TextView unitName = convertView.findViewById(R.id.unitName);
                TextView unitValue = convertView.findViewById(R.id.unitValue);
                unitValue.setVisibility(View.GONE);
                ImageView chkEnable = convertView.findViewById(R.id.chkSelected);
                chkEnable.setVisibility(View.VISIBLE);
                unitName.setText(Html.fromHtml(collections[currentCategory].get(position).getName()));

                if (position == currentUnitIndex) {
                    unitsList.setItemChecked(position, true);
                }

                itemContainer.setBackgroundColor(ContextCompat.getColor(ConvertActivity.this,
                        android.R.color.transparent));
                chkEnable.setImageResource(collections[currentCategory].get(position).isEnabled()
                        ? R.drawable.ic_check_box_white_24dp
                        : R.drawable.ic_check_box_outline_blank_white_24dp);

            } else {

                if (collections[currentCategory].get(position).isEnabled()) {
                    if (convertView == null) {
                        convertView = getLayoutInflater().inflate(R.layout.unit_listitem, parent, false);
                    }
                    View itemContainer = convertView.findViewById(R.id.unitItemContainer);
                    TextView unitName = convertView.findViewById(R.id.unitName);
                    TextView unitValue = convertView.findViewById(R.id.unitValue);
                    unitValue.setVisibility(View.VISIBLE);
                    ImageView chkEnable = convertView.findViewById(R.id.chkSelected);
                    chkEnable.setVisibility(View.GONE);
                    unitName.setText(Html.fromHtml(collections[currentCategory].get(position).getName()));

                    if (position == currentUnitIndex) {
                        unitsList.setItemChecked(position, true);
                    }

                    ViewCompat.setBackground(itemContainer,
                            ContextCompat.getDrawable(ConvertActivity.this,
                                    R.drawable.selectable_item_background));

                    double p = UnitCollection.convert(ConvertActivity.this, currentCategory,
                            currentUnitIndex, position, currentValue);
                    unitValue.setText(getFormattedValueStr(p));
                } else {
                    if (convertView == null) {
                        AbsListView.LayoutParams params
                                = new AbsListView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0);
                        convertView = new Space(ConvertActivity.this);
                        convertView.setLayoutParams(params);
                    }
                }
            }
            return convertView;
        }
    }

    public static Spanned getFormattedValueStr(double value) {
        String strValue = getValueStr(value);
        try{
            if(strValue.contains("E")){
                strValue = strValue.replace("E", " × 10<sup><small>");
                strValue += "</small></sup>";
            }
        }catch(Exception e){
            Log.d(TAG, "Error while rendering unit.", e);
        }
        return Html.fromHtml(strValue);
    }

    private static String getValueStr(double val) {
        if((Math.abs(val) > 1e6) || (Math.abs(val) < 1e-6 && Math.abs(val) > 0.0)){
            return dfExp.format(val);
        } else {
            return dfNoexp.format(val);
        }
    }

    private void doLongPressMenu(@NonNull View parentView, final int position) {
        PopupMenu menu = new PopupMenu(this, parentView, Gravity.END | Gravity.CENTER_HORIZONTAL);
        menu.getMenuInflater().inflate(R.menu.menu_long_press, menu.getMenu());
        menu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem menuItem) {
                String resultStr;
                switch (menuItem.getItemId()) {
                    case R.id.menu_copy_value:
                        resultStr = String.format("%1$s", getValueStr(UnitCollection.convert(ConvertActivity.this, currentCategory, currentUnitIndex, position, currentValue)));
                        setClipboardText(resultStr);
                        return true;
                    case R.id.menu_copy_row:
                        resultStr = String.format("%1$s %2$s = %3$s %4$s", getValueStr(currentValue),
                                collections[currentCategory].get(currentUnitIndex).getName(),
                                getValueStr(UnitCollection.convert(ConvertActivity.this, currentCategory, currentUnitIndex, position, currentValue)),
                                collections[currentCategory].get(position).getName());
                        setClipboardText(resultStr);
                        return true;
                    default:
                        break;
                }
                return false;
            }
        });
        menu.show();
    }

    private void setClipboardText(@NonNull String text) {
        ClipboardManager clipboard = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
        if (clipboard == null) {
            return;
        }
        clipboard.setPrimaryClip(ClipData.newPlainText("", text));
        Toast.makeText(ConvertActivity.this, R.string.menu_clipboard_copied, Toast.LENGTH_SHORT).show();
    }

    private void showAboutDialog() {
        new AlertDialog.Builder(this)
                .setTitle(getString(R.string.menu_about))
                .setMessage(getString(R.string.about_message))
                .setPositiveButton(R.string.ok, null)
                .create()
                .show();
    }
}

File: app/src/main/java/com/defianttech/convertme/ConvertActivity.kt
package com.defianttech.convertme

import android.content.*
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.text.Spanned
import android.util.Log
import android.view.*
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import android.widget.AdapterView.OnItemLongClickListener
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.ColorInt
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.appcompat.widget.PopupMenu
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.isVisible
import com.defianttech.convertme.NumberPadView.OnValueChangedListener
import com.defianttech.convertme.databinding.ConvertmeBinding
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import java.text.DecimalFormat
import kotlin.math.abs

/*
* Copyright (c) 2014-2022 Dmitry Brant
*/
class ConvertActivity : AppCompatActivity() {
    private lateinit var binding: ConvertmeBinding

    private lateinit var collections: List<UnitCollection>
    private lateinit var allCategoryNames: List<String>

    private var currentCategory = UnitCollection.DEFAULT_CATEGORY
    private var currentUnitIndex = UnitCollection.DEFAULT_FROM_INDEX
    private var currentValue = UnitCollection.DEFAULT_VALUE

    private lateinit var categoryMenu: PopupMenu
    private var listAdapter = UnitListAdapter()
    private var actionMode: ActionMode? = null
    private var editModeEnabled = false

    private val customUnitsLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
        if (it.resultCode == RESULT_CODE_CUSTOM_UNITS_CHANGED) {
            resetLists()
            restoreSettings()
        }
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ConvertmeBinding.inflate(layoutInflater)
        setContentView(binding.root)

        resetLists()

        setSupportActionBar(binding.mainToolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(false)
        supportActionBar?.setDisplayShowTitleEnabled(false)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            window.navigationBarColor = getColor(R.color.number_pad_background)
        }

        binding.toolbarContents.categoryToolbarContainer.setOnClickListener { categoryMenu.show() }
        categoryMenu = PopupMenu(this@ConvertActivity, binding.toolbarContents.categoryToolbarContainer)

        for ((i, name) in allCategoryNames.withIndex()) {
            categoryMenu.menu.add(0, i, 0, name)
        }

        categoryMenu.setOnMenuItemClickListener {
            currentCategory = UnitCollection.collectionIndexByName(collections,
                    allCategoryNames[it.itemId])
            if (currentUnitIndex >= collections[currentCategory].length()) {
                currentUnitIndex = 0
            }
            binding.toolbarContents.categoryText.text = it.title
            listAdapter.notifyDataSetInvalidated()
            true
        }

        binding.unitsList.adapter = listAdapter
        binding.unitsList.onItemClickListener = OnItemClickListener { _, _, position, _ ->
            if (editModeEnabled) {
                collections[currentCategory][position]
                        .isEnabled = !collections[currentCategory][position].isEnabled
            } else {
                currentUnitIndex = position
            }
            listAdapter.notifyDataSetChanged()
        }

        binding.unitsList.onItemLongClickListener = OnItemLongClickListener { _, view, position, _ ->
            if (editModeEnabled) {
                return@OnItemLongClickListener false
            }
            doLongPressMenu(view.findViewById(R.id.unitValue), position)
            true
        }

        binding.numberPad.valueChangedListener = object : OnValueChangedListener {
            override fun onValueChanged(value: String) {
                setValueFromNumberPad(value)
                listAdapter.notifyDataSetChanged()
            }
        }

        restoreSettings()

        for (name in allCategoryNames) {
            if (name == collections[currentCategory].names[0]) {
                binding.toolbarContents.categoryText.text = name
            }
        }

        binding.fabEdit.setOnClickListener {
            if (actionMode != null) {
                actionMode?.finish()
            } else {
                startSupportActionMode(EditUnitsActionModeCallback())
            }
        }

        binding.fabCustomUnits?.setOnClickListener { launchCustomUnits() }
    }

    public override fun onStop() {
        super.onStop()
        saveSettings()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_about -> {
                showAboutDialog()
                return true
            }
            R.id.menu_custom_units -> {
                launchCustomUnits()
                return true
            }
        }
        return false
    }

    private fun launchCustomUnits() {
        customUnitsLauncher.launch(Intent(this, CustomUnitsActivity::class.java))
    }

    private fun resetLists() {
        collections = UnitCollection.getInstance(this)
        allCategoryNames = UnitCollection.getAllCategoryNames(this)
        listAdapter.notifyDataSetChanged()
    }

    private fun saveSettings() {
        val editor = getPrefs(this).edit()
        editor.putInt(KEY_CURRENT_CATEGORY, currentCategory)
        editor.putInt(KEY_CURRENT_UNIT, currentUnitIndex)
        editor.putString(KEY_CURRENT_VALUE, binding.numberPad.currentValue)
        for (col in collections) {
            for (unit in col.items) {
                editor.putBoolean(unit.name, unit.isEnabled)
            }
        }
        editor.apply()
    }

    private fun restoreSettings() {
        try {
            val prefs = getPrefs(this)
            currentCategory = prefs.getInt(KEY_CURRENT_CATEGORY, UnitCollection.DEFAULT_CATEGORY)
            if (currentCategory >= collections.size) {
                currentCategory = UnitCollection.DEFAULT_CATEGORY
            }
            currentUnitIndex = prefs.getInt(KEY_CURRENT_UNIT, UnitCollection.DEFAULT_FROM_INDEX)
            if (currentUnitIndex >= collections[currentCategory].length()) {
                currentUnitIndex = 0
            }
            binding.numberPad.currentValue = prefs.getString(KEY_CURRENT_VALUE, "1")!!
            setValueFromNumberPad(binding.numberPad.currentValue)
            for (col in collections) {
                for (unit in col.items) {
                    unit.isEnabled = prefs.getBoolean(unit.name, true)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error restoring settings", e)
        }
    }

    private fun updateActionModeState() {
        binding.numberPad.isVisible = !editModeEnabled
        if (editModeEnabled) {
            binding.fabCustomUnits?.show()
        } else {
            binding.fabCustomUnits?.hide()
        }
        listAdapter.notifyDataSetInvalidated()
    }

    private fun setValueFromNumberPad(value: String) {
        currentValue = try {
            value.toDouble()
        } catch (e: NumberFormatException) {
            0.0
        }
    }

    private inner class EditUnitsActionModeCallback : ActionMode.Callback {
        @ColorInt var statusBarColor = 0

        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            actionMode = mode
            actionMode!!.title = getString(R.string.show_hide_units)
            editModeEnabled = true
            updateActionModeState()
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                statusBarColor = window.statusBarColor
                window.statusBarColor = Color.BLACK
            }
            return false
        }

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            if (item.itemId == android.R.id.home) {
                mode.finish()
                return true
            }
            return false
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                window.statusBarColor = statusBarColor
            }
            actionMode = null
            editModeEnabled = false
            updateActionModeState()
        }
    }

    private inner class UnitListAdapter : BaseAdapter() {
        override fun getCount(): Int {
            return collections[currentCategory].length()
        }

        override fun getItem(position: Int): Any {
            return collections[currentCategory][position]
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }

        override fun getItemViewType(position: Int): Int {
            return if (editModeEnabled) {
                0
            } else {
                if (collections[currentCategory][position].isEnabled) 0 else 1
            }
        }

        override fun getViewTypeCount(): Int {
            return 2
        }

        override fun getView(position: Int, convView: View?, parent: ViewGroup): View {
            var convertView = convView
            if (editModeEnabled) {
                if (convertView == null) {
                    convertView = layoutInflater.inflate(R.layout.unit_listitem, parent, false)
                }
                val itemContainer = convertView!!.findViewById<View>(R.id.unitItemContainer)
                val unitName = convertView.findViewById<TextView>(R.id.unitName)
                val unitValue = convertView.findViewById<TextView>(R.id.unitValue)
                unitValue.visibility = View.GONE
                val chkEnable = convertView.findViewById<ImageView>(R.id.chkSelected)
                chkEnable.visibility = View.VISIBLE
                unitName.text = Util.fromHtml(collections[currentCategory][position].name)
                if (position == currentUnitIndex) {
                    binding.unitsList.setItemChecked(position, true)
                }
                itemContainer.setBackgroundColor(ContextCompat.getColor(this@ConvertActivity,
                        android.R.color.transparent))
                chkEnable.setImageResource(if (collections[currentCategory][position].isEnabled) R.drawable.ic_check_box_white_24dp else R.drawable.ic_check_box_outline_blank_white_24dp)
            } else {
                if (collections[currentCategory][position].isEnabled) {
                    if (convertView == null) {
                        convertView = layoutInflater.inflate(R.layout.unit_listitem, parent, false)
                    }
                    val itemContainer = convertView!!.findViewById<View>(R.id.unitItemContainer)
                    val unitName = convertView.findViewById<TextView>(R.id.unitName)
                    val unitValue = convertView.findViewById<TextView>(R.id.unitValue)
                    unitValue.visibility = View.VISIBLE
                    val chkEnable = convertView.findViewById<ImageView>(R.id.chkSelected)
                    chkEnable.visibility = View.GONE
                    unitName.text = Util.fromHtml(collections[currentCategory][position].name)
                    if (position == currentUnitIndex) {
                        binding.unitsList.setItemChecked(position, true)
                    }
                    ViewCompat.setBackground(itemContainer,
                            ContextCompat.getDrawable(this@ConvertActivity,
                                    R.drawable.selectable_item_background))
                    val p = UnitCollection.convert(this@ConvertActivity, currentCategory,
                            currentUnitIndex, position, currentValue)
                    unitValue.text = getFormattedValueStr(p)
                } else {
                    if (convertView == null) {
                        val params = AbsListView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0)
                        convertView = View(this@ConvertActivity)
                        convertView.layoutParams = params
                    }
                }
            }
            return convertView
        }
    }

    private fun doLongPressMenu(parentView: View, position: Int) {
        val menu = PopupMenu(this, parentView, Gravity.END or Gravity.CENTER_HORIZONTAL)
        menu.menuInflater.inflate(R.menu.menu_long_press, menu.menu)
        menu.setOnMenuItemClickListener(PopupMenu.OnMenuItemClickListener {
            val resultStr: String
            when (it.itemId) {
                R.id.menu_copy_value -> {
                    resultStr = String.format("%1\$s", getValueStr(UnitCollection.convert(this@ConvertActivity, currentCategory, currentUnitIndex, position, currentValue)))
                    setClipboardText(resultStr)
                    return@OnMenuItemClickListener true
                }
                R.id.menu_copy_row -> {
                    resultStr = String.format("%1\$s %2\$s = %3\$s %4\$s", getValueStr(currentValue),
                            collections[currentCategory][currentUnitIndex].name,
                            getValueStr(UnitCollection.convert(this@ConvertActivity, currentCategory, currentUnitIndex, position, currentValue)),
                            collections[currentCategory][position].name)
                    setClipboardText(resultStr)
                    return@OnMenuItemClickListener true
                }
            }
            false
        })
        menu.show()
    }

    private fun setClipboardText(text: String) {
        val clipboard = getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
        clipboard.setPrimaryClip(ClipData.newPlainText("", text))
        Toast.makeText(this@ConvertActivity, R.string.menu_clipboard_copied, Toast.LENGTH_SHORT).show()
    }

    private fun showAboutDialog() {
        MaterialAlertDialogBuilder(this)
                .setTitle(getString(R.string.menu_about))
                .setMessage(getString(R.string.about_message))
                .setPositiveButton(R.string.ok, null)
                .show()
    }

    companion object {
        private const val TAG = "ConvertActivity"
        private const val PREFS_NAME = "ConvertMePrefs"
        private const val KEY_CURRENT_CATEGORY = "currentCategory"
        private const val KEY_CURRENT_UNIT = "currentUnitIndex"
        private const val KEY_CURRENT_VALUE = "currentValue"
        const val RESULT_CODE_CUSTOM_UNITS_CHANGED = 1
        const val INTENT_EXTRA_UNIT_ID = "extra_unit_id"
        private val dfExp = DecimalFormat("#.#######E0")
        private val dfNoexp = DecimalFormat("#.#######")

        fun getPrefs(context: Context): SharedPreferences {
            return context.getSharedPreferences(PREFS_NAME, 0)
        }

        fun getFormattedValueStr(value: Double): Spanned {
            var strValue = getValueStr(value)
            try {
                if (strValue.contains("E")) {
                    strValue = strValue.replace("E", " × 10<sup><small>")
                    strValue += "</small></sup>"
                }
            } catch (e: Exception) {
                Log.d(TAG, "Error while rendering unit.", e)
            }
            return Util.fromHtml(strValue)
        }

        private fun getValueStr(value: Double): String {
            return if (abs(value) > 1e6 || abs(value) < 1e-6 && abs(value) > 0.0) {
                dfExp.format(value)
            } else {
                dfNoexp.format(value)
            }
        }
    }
}
